### PR TITLE
[hivemind.Optimizer] ProgressTracker

### PR DESCRIPTION
**Pull request by justheuristic**
Auxiliary class that keeps track of local & global training progress, measured in epochs.
An epoch can be incremented after collaboration accumulates a said number of gradients (target_batch_size).
Similarly to pytorch LR scheduler, epoch can be incremented on a single optimizer update or many local updates.

### CODE REVIEW

**Comment by justheuristic** [comment]
Massive test results:

#5a4c56b test accuracy = 29 / 31 (too flappy, trying more lenient thresholds)
image

File: tests/test_optimizer.py [hunk 1]
```diff
@@ -170,3 +173,108 @@ def test_load_state_from_peers():
...
+    # note to a curious reader: no, you cannot reduce the timings without compromising realism or stability
+    prefix = "my_exp"
+    target_batch_size = 256
+    dht_root = hivemind.DHT(start=True)
+    barrier = mp.Barrier(parties=5)
+    delayed_start_evt = mp.Event()
+    finished_evt = mp.Event()
+    emas = mp.Array(ctypes.c_double, 5)
+
+    def worker(index: int, batch_size: int, period: float, **kwargs):
```
Comment (borzunov) [inline tests/test_optimizer.py:189]:
```suggestion
    def run_worker(index: int, batch_size: int, period: float, **kwargs):
```

nit: This makes it clear that this is a function, not an object.

File: tests/test_optimizer.py [hunk 2]
```diff
@@ -170,3 +173,108 @@ def test_load_state_from_peers():
...
+        emas[index] = tracker.performance_ema.samples_per_second
+        tracker.shutdown()
+        dht.shutdown()
+
+    # note: we use processes here because RSASignatureValidator inside trackers uses process-wide RSA keys
+    worker1 = mp.Process(target=worker, kwargs=dict(index=1, batch_size=12, period=0.6))
+    worker2 = mp.Process(target=worker, kwargs=dict(index=2, batch_size=16, period=0.5))
+    worker3 = mp.Process(target=worker, kwargs=dict(index=3, batch_size=24, period=0.4))
+    worker4 = mp.Process(target=worker, kwargs=dict(index=4, batch_size=64, period=0.4))
+    workers = worker1, worker2, worker3, worker4
```
Comment (borzunov) [inline tests/test_optimizer.py:235]:
```suggestion
    workers = [
        mp.Process(target=worker, kwargs=dict(index=1, batch_size=12, period=0.6)),
        mp.Process(target=worker, kwargs=dict(index=2, batch_size=16, period=0.5)),
        mp.Process(target=worker, kwargs=dict(index=3, batch_size=24, period=0.4)),
        mp.Process(target=worker, kwargs=dict(index=4, batch_size=64, period=0.4)),
    ]
```

nit: No need for separate variables for each worker.

File: hivemind/optim/experimental/progress_tracker.py [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+        self._local_public_key = signature_validator.local_public_key
+        dht.add_validators([SchemaValidator(TrainingProgressSchema, prefix=prefix), signature_validator])
+
+        # report the collaboration progress periodically or in background
+        self.local_progress = self._get_local_progress(local_epoch=0, samples_accumulated=0)
+        metadata, _expiration = self.dht.get(self.training_progress_key, latest=True) or (None, -float("inf"))
+        self.global_progress = self._parse_swarm_progress_data(metadata)
+        self.lock_global_progress, self.global_state_updated = Lock(), Event()
+        self.lock_local_progress, self.should_report_progress = Lock(), Event()
+        self.shutdown_triggered, self.shutdown_complete = Event(), Event()
```
Comment (borzunov) [inline hivemind/optim/experimental/progress_tracker.py:117]:
```suggestion
        self.lock_global_progress, self.global_state_updated = threading.Lock(), threading.Event()
        self.lock_local_progress, self.should_report_progress = threading.Lock(), threading.Event()
        self.shutdown_triggered, self.shutdown_complete = threading.Event(), threading.Event()
```

Let's use `threading` explicitly: currently, I needed to scroll to the imports to find out whether these are threading, mp, or asyncio locks/events (all these types are widely used across our code).

File: hivemind/optim/experimental/progress_tracker.py [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+        target_batch_size: int,
+        *,
+        client_mode: Optional[bool] = None,
+        min_refresh_period: float = 0.5,
+        max_refresh_period: float = 30,
+        default_refresh_period: float = 3,
+        expected_drift_peers: float = 3,
+        expected_drift_rate: float = 0.2,
+        performance_ema_alpha: float = 0.1,
+        metadata_expiration: float = 30.0,
```
Comment (borzunov) [inline hivemind/optim/experimental/progress_tracker.py:91]: Are these defaults reasonable for all configurations?

File: hivemind/optim/experimental/progress_tracker.py [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+    >>>     accumulate_gradients(batch_size=32)
+    >>>     local_samples += 32
+    >>>     tracker.report_local_progress(local_epoch, local_samples)
+    >>>     if local_epoch < tracker.global_progress.epoch:
+    >>>         download_state_from_peers()  # if peer is out of sync, synchronize it with the swarm
+    >>>     if tracker.accumulated_enough_samples:
+    >>>         with tracker.pause_updates():
+    >>>             aggregate_gradients_with_peers()
+    >>>             update_model_parameters()
+    >>>             local_epoch = tracker.increment_epoch(local_epoch + 1)
```
Comment (borzunov) [inline hivemind/optim/experimental/progress_tracker.py:74]:
```suggestion
    >>>             local_epoch = tracker.update_epoch(local_epoch + 1)
```

I'd expect a function `increment_smth(x)` to calculate `x + 1` by itself.

File: hivemind/optim/experimental/progress_tracker.py [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+    Example:
+
+    >>> tracker = ProgressTracker(hivemind.DHT(...), prefix="my_experiment_with_several_peers", target_batch_size=100)
+    >>> local_epoch, local_samples = 0, 0
+    >>> while True:
+    >>>     accumulate_gradients(batch_size=32)
+    >>>     local_samples += 32
+    >>>     tracker.report_local_progress(local_epoch, local_samples)
+    >>>     if local_epoch < tracker.global_progress.epoch:
+    >>>         download_state_from_peers()  # if peer is out of sync, synchronize it with the swarm
```
Comment (borzunov) [inline hivemind/optim/experimental/progress_tracker.py:69]: Do I understand correctly that this code is similar to what is written (or will be written) in hivemind.Optimizer?

File: hivemind/optim/experimental/progress_tracker.py [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+        self.performance_ema = PerformanceEMA(alpha=performance_ema_alpha)
+        self.metadata_expiration = metadata_expiration
+
+        signature_validator = RSASignatureValidator(private_key=RSAPrivateKey())
+        self._local_public_key = signature_validator.local_public_key
+        dht.add_validators([SchemaValidator(TrainingProgressSchema, prefix=prefix), signature_validator])
+
+        # report the collaboration progress periodically or in background
+        self.local_progress = self._get_local_progress(local_epoch=0, samples_accumulated=0)
+        metadata, _expiration = self.dht.get(self.training_progress_key, latest=True) or (None, -float("inf"))
```
Comment (borzunov) [inline hivemind/optim/experimental/progress_tracker.py:113]: suggestion (here and further):

```suggestion
        metadata, _expiration = self.dht.get(self.training_progress_key, latest=True) or (None, -np.inf)
```

`np.inf`/`math.inf` also give the native Python float, but is shorter and don't look as a weird string conversion :)

File: hivemind/optim/experimental/progress_tracker.py [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+            else:
+                logger.debug("Resetting performance timestamp to current time (progress was reset)")
+                self.performance_ema.reset_timer()
+            self.local_progress = self._get_local_progress(local_epoch, samples_accumulated)
+            self.should_report_progress.set()
+
+    @contextlib.contextmanager
+    def pause_updates(self):
+        """Temporarily stop progress tracker from updating global training state"""
+        with self.lock_global_progress, self.performance_ema.pause():
```
Comment (borzunov) [inline hivemind/optim/experimental/progress_tracker.py:170]: (may be out of scope of this PR)

I'm surprised that `performance_ema.pause()` resets the timer instead of actually pausing it.

File: hivemind/optim/experimental/progress_tracker.py [hunk 9]
```diff
@@ -0,0 +1,326 @@
...
+
+                async with enter_asynchronously(self.lock_global_progress):
+                    progress_entry = await self.dht.get(self.training_progress_key, latest=True, return_future=True)
+                    metadata = progress_entry.value if isinstance(progress_entry, ValueWithExpiration) else None
+                    self.global_progress = self._parse_swarm_progress_data(metadata)
+        finally:
+            logger.log(self.status_loglevel, f"No longer fetching {self.training_progress_key}.")
+
+    def _parse_swarm_progress_data(
+        self, metadata: Optional[Dict[BytesWithPublicKey, Optional[LocalTrainingProgress]]]
```
Comment (borzunov) [inline hivemind/optim/experimental/progress_tracker.py:242]:
```suggestion
        self, metadata: TrainingProgressSchema
```

File: hivemind/optim/experimental/progress_tracker.py [hunk 10]
```diff
@@ -0,0 +1,326 @@
...
+
+        for peer in valid_peer_entries:
+            total_samples_per_second += peer.samples_per_second
+            if peer.epoch == global_epoch:
+                total_samples_accumulated += peer.samples_accumulated
+                estimated_current_samples += (
+                    peer.samples_accumulated + max(0.0, current_time - peer.time) * peer.samples_per_second
+                )
+            # note: we deliberately count only valid peers for samples_accumulated, but all peers for performance;
+            # the rationale behind this is that outdated peers will synchronize and begin contributing shortly.
```
Comment (borzunov) [inline hivemind/optim/experimental/progress_tracker.py:287]: Are outdated peers really more common than disconnecting peers? If not, maybe exclude invalid peers from the perf estimate as well?

**Review by borzunov — CHANGES_REQUESTED** [review]

File: hivemind/optim/experimental/progress_tracker.py (hunk 4 above)
Comment (justheuristic) [reply hivemind/optim/experimental/progress_tracker.py:91]: yes, in all we've tested so far

File: hivemind/optim/experimental/progress_tracker.py (hunk 6 above)
Comment (justheuristic) [reply hivemind/optim/experimental/progress_tracker.py:69]: hivemind.Optimizer is significantly more complicated, but this is one of the special cases

File: hivemind/optim/experimental/progress_tracker.py (hunk 7 above)
Comment (justheuristic) [reply hivemind/optim/experimental/progress_tracker.py:113]: Keeping the original version for the sake of consistency.

File: hivemind/optim/experimental/progress_tracker.py [hunk 11]
```diff
@@ -0,0 +1,326 @@
...
+        self.training_progress_key = f"{self.prefix}_progress"
+        self.target_batch_size = target_batch_size
+        self.min_refresh_period, self.max_refresh_period = min_refresh_period, max_refresh_period
+        self.default_refresh_period = default_refresh_period
+        self.expected_drift_peers, self.expected_drift_rate = expected_drift_peers, expected_drift_rate
+        self.status_loglevel = status_loglevel
+        self.performance_ema = PerformanceEMA(alpha=performance_ema_alpha)
+        self.metadata_expiration = metadata_expiration
+
+        signature_validator = RSASignatureValidator(private_key=RSAPrivateKey())
```
[the remark this answers is not in the document]
Comment (justheuristic) [reply hivemind/optim/experimental/progress_tracker.py]: good point. added that

File: tests/test_optimizer.py [hunk 12]
```diff
@@ -170,3 +174,109 @@ def test_load_state_from_peers():
...
+    finished_evt.set()
+    for worker in workers:
+        worker.join()
+
+    tracker.shutdown()
+    dht_root.shutdown()
+    assert not tracker.is_alive()
+
+    mean_step_time = sum(step_time_deltas) / len(step_time_deltas)
+    for i in (0, 1, 5):
```
Comment (borzunov) [inline tests/test_optimizer.py:277]:
```suggestion
    for i in (0, 1, 5):  # Without the 4th worker (the fastest one)
```

File: tests/test_optimizer.py [hunk 13]
```diff
@@ -170,3 +174,109 @@ def test_load_state_from_peers():
...
+        assert 1.05 * mean_step_time < step_time_deltas[i] < 2.0 * mean_step_time
+    for i in (2, 3, 4):
```
Comment (borzunov) [inline tests/test_optimizer.py:279]:
```suggestion
    for i in (2, 3, 4):  # With the 4th worker
```

**Review by borzunov — APPROVED** [review]

File: hivemind/optim/experimental/progress_tracker.py [hunk 14]
```diff
@@ -0,0 +1,321 @@
...
+        client_mode: Optional[bool] = None,
+        min_refresh_period: float = 0.5,
+        max_refresh_period: float = 30,
+        default_refresh_period: float = 3,
+        expected_drift_peers: float = 3,
+        expected_drift_rate: float = 0.2,
+        performance_ema_alpha: float = 0.1,
+        metadata_expiration: float = 30.0,
+        status_loglevel: int = logging.DEBUG,
+        private_key: PrivateKey = None,
```
Comment (borzunov) [inline hivemind/optim/experimental/progress_tracker.py:93]:
```suggestion
        private_key: Optional[RSAPrivateKey] = None,
```

File: tests/test_optimizer.py [hunk 15]
```diff
@@ -170,3 +174,109 @@ def test_load_state_from_peers():
...
+
+                if index == 4 and local_epoch >= 5:
+                    time.sleep(0.5)
+                    break
+
+        emas[index] = tracker.performance_ema.samples_per_second
+        tracker.shutdown()
+        dht.shutdown()
+
+    # note: we use processes here because RSASignatureValidator inside trackers uses process-wide RSA keys
```
[the remark this answers is not in the document]
Comment (borzunov) [reply tests/test_optimizer.py:232]:
```suggestion
```

Comment (borzunov) [inline tests/test_optimizer.py:232]:
```suggestion
```
